### ISSUE TITLE
vNext bundle size requirements and tooling support

### ISSUE THREAD

**Issue by jdhuntington** [microsoft/fluentui#17678]
## Problem

One important aspect to supporting fluentui's users is to ensure the total number of bytes in a compiled bundle is as small as possible. However, this necessarily comes at a cost of features, capabilities, and surface area. 
Two specific possible problems are easy to visualize around bundle size:
1. "as small as possible" bundle size becomes the high order bit, and the team is hamstrung from making any progress because any changes come at the cost of increased bytes.
2. unchecked or "lip service" attention to bundle size means the bundle could grow unbounded, hurting consumers of fluentui by making their payloads too large.

## Appetite

About 2 weeks. A meaningful portion of the solution will take place in meetings, so the 2 weeks is not necessarily linear.

## Solution

There are 2 distinct portions of the solution, and they can be carried out in parallel.

### Determine requirements

We will meet with likely internal consumers and determine their ultimate goals for bundle size impact from components. We will also survey 3-6 popular open source react component libraries and see how their bundle size stacks up for a set of scenarios. Based on the results from this research, we will create a matrix of goals. One set of targets is _ideal_ and should be maintained, another (slightly more relaxed) set of targets will be the "do not pass" line that we must not exceed.

### Tooling support

Once defined, we will build tooling into the local development environment as well as the official build to indicate warnings or outright failures. This work can start on day 0 with arbitrary goals that we can later refine once the requirement definition is complete.

## Risks

- Not unique to this solution, but any bundle size requirements will likely pose productivity issues for developers building components. It would be ideal to malloc bundle size impact to a developer before starting a feature so that slack wouldn't be taken up by other features between a feature starting and being merged.

## Out of scope (No-gos)

- For the initial implementation, CI will point out specific errors or warnings, but will not host a user friendly bundle size investigation tool. It will be up to developers to run that from their own environment.
- Historical tracking of bundle size measures.

**Comment by jdhuntington** [comment]
Rough draft of requirements doc: https://hackmd.io/HlQnPX9_SK-Wu5Qp_rYKnQ?view

**Comment by ecraig12345** [comment]
@layershifter and @miroslavstastny implemented tooling for this. Closing this for now but it's still here for reference if needed.
